google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,698 @@
+            next_page_token = response.get('nextPageToken')
+            yield response
+
+            if not next_page_token:
+                break
+
+    def execute_query(self, verb, verb_arguments):
+        """Executes query (ex. get) via a dedicated http object.
+
+        Args:
+          verb (str): Method to execute on the component (ex. get, list).
+          verb_arguments (dict): key-value pairs to be passed to _BuildRequest.
+
+        Returns:
+          dict: Service Response.
+        """
+        request = self._build_request(verb, verb_arguments)
+        return self._execute(request)
+
+    @retry(retry_on_exception=retryable_exceptions.is_retryable_exception,
+           wait_exponential_multiplier=1000, wait_exponential_max=10000,
+           stop_max_attempt_number=5)
+    def _execute(self, request):
+        """Run execute with retries and rate limiting.
+
+        Args:
+          request (object): The HttpRequest object to execute.
+
+        Returns:
+          dict: The response from the API.
+        """
+        if self._rate_limiter:
+            # Since the ratelimiter library only exposes a context manager
+            # interface the code has to be duplicated to handle the case where
+            # no rate limiter is defined.
+            with self._rate_limiter:
+                return request.execute(http=self.http,
+                                       num_retries=self._num_retries)
+        else:
+            return request.execute(http=self.http,
+                                   num_retries=self._num_retries)
+# pylint: enable=too-many-instance-attributes, too-many-arguments
+
+
+class ListQueryMixin(object):
+    """Mixin that implements Paged List query."""
+
+    def list(self, resource=None, fields=None, max_results=None, verb='list',
+             **kwargs):
+        """List GCP entities of a given resource.
+
+        Args:
+          resource (str): The id of the resource to query.
+          fields (str): Fields to include in the response - partial response.
+          max_results (int): Number of entries to include per page.
+          verb (str): The method to call on the API.
+          kwargs (dict): Optional additional arguments to pass to the query.
+
+        Yields:
+          dict: GCE response.
+        """
+        assert isinstance(self, GCPRepository)
+
+        arguments = {'fields': fields,
+                     self._max_results_field: max_results}
+
+        if self._list_key_field and resource:
+            arguments[self._list_key_field] = resource
+
+        if kwargs:
+            arguments.update(kwargs)
+
+        try:
+            for resp in self.execute_paged_query(verb=verb,
+                                                 verb_arguments=arguments):
+                yield resp
+        except api_errors.PaginationNotSupportedError:
+            # Some API list() methods are not actually paginated.
+            del arguments[self._max_results_field]
```

Comment:
I think this logic should be moved into a decorator  - let's know raise an exception for the case we know how to deal with. The decorator would check if the method supports pagination and del max_results_field.